spec/helpers/http_mock.lua

```diff
@@ -0,0 +1,176 @@
+local helpers = require "spec.helpers"
+
+local pairs = pairs
+local ipairs = ipairs
+local type = type
+local setmetatable = setmetatable
+
+local modules = {
+  require "spec.helpers.http_mock.nginx_instance",
+  require "spec.helpers.http_mock.asserts",
+  require "spec.helpers.http_mock.debug_port",
+  require "spec.helpers.http_mock.clients",
+}
+
+local http_mock = {}
+
+-- the http_mock is a too large module, so we split it into several modules
+for _, module in ipairs(modules) do
+  for k, v in pairs(module) do
+    http_mock[k] = v
+  end
+end
+
+local http_mock_MT = { __index = http_mock, __gc = http_mock.stop }
+
+
+-- TODO: make default_mocking the same to the `mock_upstream`
+local default_mocking = {
+  ["/"] = {
+    access = [[
+      ngx.req.set_header("X-Test", "test")
+      ngx.print("ok")
+      ngx.exit(200)
+    ]],
+  },
+}
+
+local function default_field(tbl, key, default)
+  if tbl[key] == nil then
+    tbl[key] = default
+  end
+end
+
+-- create a mock instance which represents a HTTP mocking server
+-- @param listens: the listen directive of the mock server, default to "0.0.0.0:8000"
+-- @param code: the code of the mock server, default to a simple response.
+-- @param opts: options for the mock server, left it empty to use the defaults
+-- @return: a mock instance
+-- @usage
+-- local mock = http_mock.new(8000, [[
+--   ngx.req.set_header("X-Test", "test")
+--   ngx.print("hello world")
+-- ]],  {
+--   prefix = "mockserver",
+--   record_opts = {
+--     resp = true,
+--     resp_body = true,
+--   },
+--   tls = true,
+-- })
+--
+-- mock:start()
+-- local client = mock:get_client() -- get a client to access the mocking port
+-- local res = assert(client:send({}))
+-- assert.response(res).has.status(200)
+-- assert.response(res).has.header("X-Test", "test")
+-- assert.response(res).has.body("hello world")
+-- mock.eventually:has_response(function(resp)
+--   assert.same(resp.body, "hello world")
+-- end)
+-- mock:wait_until_no_request() -- wait until all the requests are finished
+-- mock:clean() -- clean the records
+-- client:send({})
+-- client:send({})
+-- local records = mock:get_record() -- get all the records of HTTP sessions
+-- mock:stop()
+-- 
+-- listens can be a number, which will be used as the port of the mock server;
+-- or a string, which will be used as the param of listen directive of the mock server;
+-- or a table represents multiple listen ports.
+-- if the port is not specified, a random port will be used.
```

Comment:
How can tests determine the port number that was chosen?

Reply:
The random port is used only when the user does not care which port is used. In this case, the user may call `mock:get_client()` to get a client to send requests to the mocking server.

Reply:
The "listens" could be an array of strings, it may also use server HTTP and HTTPS.
Anyway, I'm exposing a specific port via the `mock.get_default_port()`.

Reply:
@hanshuebner Just added. Please refresh to see the change.